### PR TITLE
Add AWS Lambda CloudFront Event origin fields.

### PR DESCRIPTION
**Pull request by ohookins**
This is my first foray into TypeScript so please let me know if anything needs to be changed. The existing tests already contained the whole schema but there were a number of missing interfaces from the definition. It's perhaps too permissive (several optional fields) since the `CloudFrontRequestEvent` is used in both a request event and response event, but in the response event the `origin` field is not present. To fix would require a bit of refactoring and more duplication.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html
- [x] Increase the version number in the header if appropriate.
- ~[] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

### CODE REVIEW

File: types/aws-lambda/index.d.ts
Comment (simonbuchan) [inline types/aws-lambda/index.d.ts]: Revert this, the idea is these are definitions for the node 8.10 lambda runtime. Yeah, it's confusing when external services add features, but at least there's some version to tie this to.

File: types/aws-lambda/index.d.ts
Comment (simonbuchan) [inline types/aws-lambda/index.d.ts]: Since it's declared you must have exactly one of these, a stronger definition would be:

```ts
// the never fields are required both to require only one field is set and
// to allow reading if either field exists.
type CloudFrontOrigin =
    | { s3: CloudFrontS3Origin, custom?: never }
    | { custom: CloudFrontCustomOrigin, s3?: never };
```

but I'm not sure it's worth the extra complexity. If more fields are added in the future, this type gets ugly real fast.

File: types/aws-lambda/index.d.ts
Comment (simonbuchan) [inline types/aws-lambda/index.d.ts]: The docs were a little confusing here, so just to be clear, it's possible that this is not provided by cloudfront, e.g. that your handler should test if this is set before reading it?

**Review by simonbuchan — CHANGES_REQUESTED** [review]
Sorry for the delay reviewing this, looks good!
The version number shouldn't be changed, and if you could please add at least a quick couple of lines into the test file (the `Add or edit tests to reflect the change. (Run with ``npm test``)` checklist item) showing reading a value out and setting this field, it doesn't need to be exhaustive, that would be great.

File: types/aws-lambda/index.d.ts
Comment (ohookins) [reply types/aws-lambda/index.d.ts]: Will do.

File: types/aws-lambda/index.d.ts
Comment (ohookins) [reply types/aws-lambda/index.d.ts]: Yes, that solves the problem of the mutually exclusive options but doesn't help with the reuse of the request object in the response, which does not carry the origin field.

I like the idea of enforcing the mutual exclusivity, so let me know if you don't want this added.

File: types/aws-lambda/index.d.ts
Comment (ohookins) [reply types/aws-lambda/index.d.ts]: I'm also not 100% sure but I will check for some more API specifications and comment back.

File: types/aws-lambda/index.d.ts
Comment (simonbuchan) [reply types/aws-lambda/index.d.ts]: If you have code that uses this and it works, sure, go ahead!

The docs seem to say you can modify the origin on inbound requests, though?

> In the CloudFront origin request event, the origin element in the event structure contains information about the Amazon S3 or custom origin that the request would be routed to, based on the path pattern. You can update the values in the origin object to route a request to a different origin. The origin doesn't need to be defined in the distribution, and you can replace an Amazon S3 origin object with a custom origin object and vice versa.

I suppose you could `interface CloudFrontInboundRequest extends CloudFrontRequest { origin?: CloudFrontOrigin; }`, but that's not just more complex a definition, but adding more types for the user to think about and has back-compatibility issues - I'd say that one isn't worth it.

**Comment by ohookins** [comment]
Yes, you can modify the origin on inbound requests (this is actually our
main use case), but still only one of a custom origin or S3 origin should
be permitted. Unfortunately the "example" event from AWS contains both the
custom origin and S3 origin (I guess for demonstrative purposes) but
according to their documentation, only one would be allowed so it is
technically invalid. I'm trying to source an official schema so we could
have a better reference.

**Comment by simonbuchan** [comment]
Yeah, a formal event schema spec would be nice, but I don't think they have one for lambda events (let me know if you find one!). For the service API calls there's the service model json files you can see in the various SDK sources, but the types are only sometimes related to lambda types, and they don't have the detail we're talking about here :(

**Comment by ohookins** [comment]
@simonbuchan might need some assistance with this. I'm not familiar with the way these types are tested, but the way I've constructed the tests at least it is not failing where it should. Any suggestions for a better way to do this?

File: types/aws-lambda/aws-lambda-tests.ts
Comment (simonbuchan) [inline types/aws-lambda/aws-lambda-tests.ts]: Not sure what this and `.custom` below is checking for? If these were testing that the `.s3` was not accessible inside the `if (request.origin.custom)` due to flow analysis, that would make sense, but at the moment this seems to be some other error.

**Review by simonbuchan — APPROVED** [review]
👌 Thanks for your work and feedback here! I'm not 100% parts of the tests, but the whole file is a mess overall at the moment anyway so I wouldn't worry too much about getting it "right" as it's not clear what that even means for this package.

File: types/aws-lambda/aws-lambda-tests.ts
Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: Ahh, it's failing the tests, which is what I would be expecting. Try moving these inside the `if`s above each: e.g. `.s3` is not accessible if `.custom` is, and vice versa.

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]: Not sure if this is what you meant:
```
    let request = event.Records[0].cf.request;
    if (request.origin && request.origin.custom) {
        request.origin.custom.domainName;
        request.origin.custom.domainName = "example2.com";

        // $ExpectError
        request.origin.s3;
    }
```
Apparently this found no error.

Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: Try reading or writing to it.

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]: OK, maybe that was my misunderstanding of how these tests are supposed to work. Looking at them I assumed that simply referencing an object was enough to assert something.

Anyway:
```
    let request = event.Records[0].cf.request;
    if (request.origin && request.origin.custom) {
        request.origin.custom.domainName;
        request.origin.custom.domainName = "example2.com";

        // $ExpectError
        const s3 = request.origin.s3;

        // $ExpectError
        request.origin.s3 = s3;
    }
```
Sadly neither of these last two assignments fail.

Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: Sorry, I wasn't too clear, [this playground link](http://www.typescriptlang.org/play/#src=type%20Y%20%3D%20%7B%20c%3A%20number%20%7D%3B%0D%0Atype%20X%20%3D%0D%0A%20%20%20%20%7C%20%7B%20a%3A%20Y%2C%20b%3F%3A%20never%20%7D%0D%0A%20%20%20%20%7C%20%7B%20a%3F%3A%20never%2C%20b%3A%20Y%20%7D%3B%0D%0A%0D%0Aconst%20x1%3A%20X%20%3D%20%7B%20a%3A%20123%20%7D%3B%0D%0Aconst%20x2%3A%20X%20%3D%20%7B%20b%3A%20456%20%7D%3B%0D%0Alet%20y%3A%20Y%20%3D%20%7B%20c%3A%20789%20%7D%3B%0D%0A%0D%0Aif%20(x1.a)%20%7B%0D%0A%20%20%20%20x1.a%20%3D%20y%3B%20%2F%2F%20OK!%0D%0A%20%20%20%20y%20%3D%20x1.a%3B%20%2F%2F%20OK!%0D%0A%0D%0A%20%20%20%20x1.b%3B%20%2F%2F%20no%20error!%20reads%20undefined%0D%0A%20%20%20%20x1.b%20%3D%20y%3B%20%2F%2F%20error%2C%20assign%20Y%20to%20undefined%0D%0A%20%20%20%20y%20%3D%20x1.b%3B%20%2F%2F%20error%2C%20assign%20undefined%20to%20Y%0D%0A%7D) shows what I mean.

Basically `$ExpectError` just makes dtslint fail the test if typescript does *not* generate an error on the next line (you can also check the message with `$ExpectError message`)

Your code is not failing because it simply types the `const s3` as `undefined`, since as the property is optional TS adds `undefined` to the type, which is a bit annoying 😢

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]: Well, despite your extensive help I still can't get it working. The code even looks identical to your playground example. Sorry for all the time this is taking!

Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: Ah, just remembered that CI was failing when you opened this branch due to `react-table-filter`, try rebasing/merging up to current master so you're not getting confounded by that!

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]: Well, it's also failing locally, so...

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]: Seems like it's up to date with master as well.

Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: What's the error you're getting?

Comment (ohookins) [reply types/aws-lambda/aws-lambda-tests.ts]:
```
ERROR: 870:1  expect  Expected an error on this line, but found none.
ERROR: 873:1  expect  Expected an error on this line, but found none.
ERROR: 896:1  expect  Expected an error on this line, but found none.
ERROR: 899:1  expect  Expected an error on this line, but found none.
```
These are the exact lines that have the assertion `// $ExpectError`

Comment (simonbuchan) [reply types/aws-lambda/aws-lambda-tests.ts]: OK, did some playing around and it looks like a TS limitation. The reason you're seeing no errors is that you are assigning to the `never` field before reading, so TS just updates the type. For some reason this seems to cause it to merge the types so both fields are valid. 😵

At least these works, in this order:

```ts
// $ExpectError
customOrigin = request.origin.custom;

// $ExpectError
request.origin.custom.domainName = "blah";

// $ExpectError
request.origin = {
   custom: customOrigin,
   s3: s3Origin,
};
```

You will need to merge upsteam master for the CI to work though.

**Comment by ohookins** [comment]
Lint/test works now locally, although the reason the second line assertion fails is because the parent object could be undefined (which VSCode warns me about) rather than confirming the type constraint. Maybe that's enough?

Have also rebased against upstream/master after I figured out I wasn't in the upstream repo ;)

**Comment by simonbuchan** [comment]
Yeah, the problem is you can only be as good as typescript lets you be, and it doesn't have a real way to represent "not these properties", this is just the closest you can't get (as far as I know)

And the upstream / origin thing always gets me too 😉

**Comment by ohookins** [comment]
Thought I'd done something wrong but master is currently failing the build as well.

**Comment by ohookins** [comment]
Looks like master is back to normal, and I've rebased and pushed again and everything is passing. Good to go?

**Comment by simonbuchan** [comment]
👍 on my end! I approved the changes already, so once the bot updates the labels a DT maintainer will be around to merge this within a day, generally.

Thanks for keeping on working though all this, generally it's not this much trouble!

**Comment by ohookins** [comment]
Thanks for all your help, hopefully I can contribute again in future!

**Pull request merged**
